internal/compiler/lib.rs

```diff
@@ -171,17 +171,20 @@ pub async fn compile_syntax_node(
         &type_registry,
     );
 
-    if let Some((_, node)) = &*doc.root_component.child_insertion_point.borrow() {
+    let root_component = doc.determine_root_component(&mut diagnostics);
+
+    if let Some((_, node)) = &*root_component.child_insertion_point.borrow() {
         diagnostics
             .push_error("@children placeholder not allowed in the final component".into(), node)
     }
 
     if !diagnostics.has_error() {
         // FIXME: ideally we would be able to run more passes, but currently we panic because invariant are not met.
-        passes::run_passes(&doc, &mut diagnostics, &mut loader, &compiler_config).await;
+        passes::run_passes(&root_component, &doc, &mut diagnostics, &mut loader, &compiler_config)
+            .await;
     }
 
     diagnostics.all_loaded_files = loader.all_files().cloned().collect();
 
-    (doc, diagnostics)
+    (root_component, doc, diagnostics)
```

Comment:
Why do we return a root_component here instead of just using Document::root_component?

Reply:
Besides feeling like a natural change, this also makes it easy to issue the warning only for the true root component, instead of for every Slint file. But if you prefer then I can try to solve that differently and restore root_component in Document.